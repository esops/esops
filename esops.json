 [
   "github:sartaj/esops-ts#2019-10-18-8 --subdir ts-node",
   "github:sartaj/esops-ts#2019-10-18-8 --subdir ts-ramda",
   "./package.extend.json -o ./package.json --merge"
 ]