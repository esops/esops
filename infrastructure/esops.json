{
  "stack": ["github:sartaj/typescript-node-stack/stack#0.1.0", "./"],
  "destination": "../"
}
