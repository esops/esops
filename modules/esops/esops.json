<pipe-me />
<target-react-native-web module={'@my/app-navigation'} />
<semantic-design />