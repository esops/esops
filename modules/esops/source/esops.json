[
  "npm:sartaj/typescript-target-stack",
  "http:",
  "https:",
  "git:github.com:sartaj/typescript-stack.git",
  "github:sartaj/typescript-stack"
]
