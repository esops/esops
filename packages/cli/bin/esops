#!/usr/bin/env node

require('@esops/language-babel/node')

// run the CLI with the current process arguments
require('../')(process.argv)
