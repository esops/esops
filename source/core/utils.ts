import {is} from 'ramda'
import {Option, WithProps, Options, OptionsWithProps} from './types'
import {isString} from '../utils/sync'

export const isWithProps = opt =>
  Array.isArray(opt) &&
  opt.length === 2 &&
  typeof opt[1] === 'string' &&
  typeof opt[2] === 'object'

export const getPath = (opt: Option) =>
  is(String, opt)
    ? opt
    : is(String, opt[0])
    ? opt[0]
    : new TypeError(`${JSON.stringify(opt)} is not an object`)

export const createDefaultOpt = (path: string): WithProps => [path, {}]

export const convertAllOptionsToHaveProps = (opts: Options): OptionsWithProps =>
  isString(opts)
    ? [createDefaultOpt(opts)]
    : opts.map((opt: Option) => (isString(opt) ? createDefaultOpt(opt) : opt))

export const isValidOpts = opts => is(String, opts) || is(Array, opts)
