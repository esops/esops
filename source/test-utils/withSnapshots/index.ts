import {curry} from 'ramda'
import * as path from 'path'
import * as fs from 'fs'
import * as riteway from 'riteway'

const updateSnapshot = (name, snapshotPath, actual) => {
  fs.writeFileSync(snapshotPath, actual)
  console.log(`snapshot ${name} has been updated`)
  return {
    expected: actual,
    actual
  }
}

const toMatchSnapshot = curry((snapshotDir, name, contents) => {
  const shouldUpdateSnapshots = process.env.UPDATE_SNAPSHOTS
  const snapshotPath = path.join(snapshotDir, `${name}.snap`)

  if (!fs.existsSync(snapshotDir)) fs.mkdirSync(snapshotDir)

  const actual =
    typeof contents !== 'string' ? JSON.stringify(contents, null, 2) : contents

  if (!fs.existsSync(snapshotPath)) {
    if (shouldUpdateSnapshots) return updateSnapshot(name, snapshotPath, actual)
    else
      throw new Error(
        'Snapshot does not exist yet. Run with UPDATE_SNAPSHOTS=1 to update snapshots.'
      )
  }

  const expected = fs.readFileSync(snapshotPath, {encoding: 'utf-8'})

  if (actual !== expected) {
    if (shouldUpdateSnapshots) {
      return updateSnapshot(name, snapshotPath, actual)
    } else {
      return {
        actual,
        expected
      }
    }
  } else {
    return {
      actual,
      expected
    }
  }
})

function hashCode(str) {
  return str
    .split('')
    .reduce(
      (prevHash, currVal) =>
        ((prevHash << 5) - prevHash + currVal.charCodeAt(0)) | 0,
      0
    )
}

export const withSnapshots = curry(
  (cwd: string | null, dirname: string | null, describeMessage, callback) => {
    const dirPath: string = cwd || process.cwd()
    const snapshotDirname: string = dirname || 'snapshots'
    const snapshotDir = path.join(dirPath, snapshotDirname)
    const matchByName = toMatchSnapshot(snapshotDir)

    const describeFunc = riteway.describe
    if (typeof describeFunc === 'function') {
      describeFunc(describeMessage, async assert => {
        const assertSnap = ({given, should, snap}) => {
          const name = `${describeMessage}-${hashCode(
            describeMessage + given + should
          )}`

          assert({
            given,
            should,
            ...matchByName(name, snap)
          })
        }
        callback(assert, assertSnap)
      })
    }
  }
)
