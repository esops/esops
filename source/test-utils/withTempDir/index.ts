import * as path from 'path'
import * as rimraf from 'rimraf'
import fs from '../../drivers/fs'
import {curry} from 'ramda'

export const withTempDir = curry(async (dir, initialFiles, callback) => {
  const tempPath = path.join(dir, '/.tmp/')

  async function before() {
    rimraf.sync(tempPath, fs)
    if (!fs.existsSync(tempPath)) fs.mkdirSync(tempPath)
    if (initialFiles) fs.forceCopy(initialFiles, tempPath)
  }

  async function after() {
    rimraf.sync(tempPath, fs)
  }
  await before()
  await callback(tempPath)
  await after()
})
