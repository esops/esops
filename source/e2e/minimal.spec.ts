import * as path from 'path'
import * as spawn from 'await-spawn'

import {Try} from 'riteway'
import {withSnapshots} from '../../test-utils/withSnapshots'
import {withTempDir} from '../../test-utils/withTempDir'
import {
  getSortedFilePaths,
  getFileContents,
  getJsonContents,
  cleanErrorString
} from '../../test-utils/fs-utils'

import esops from '../run/index'
import {MOCK_STACKS} from '../core/examples'

const describe = withSnapshots(__dirname, 'snapshots-minimal')

describe('esops() minimal features', async (assert, assertSnap) => {
  await withTempDir(__dirname, MOCK_STACKS['basic-gitignore'], async cwd => {
    await esops({cwd})

    await assertSnap({
      given: 'stack with esops.json and fs path',
      should: 'generate basic template in cwd',
      snap: getSortedFilePaths(cwd)
    })

    await assertSnap({
      given: 'an included gitignore',
      should: 'have updated gitignore with generated file paths',
      snap: getFileContents(path.join(cwd, '.gitignore'))
    })
  })

  await withTempDir(__dirname, MOCK_STACKS['basic-package-json'], async cwd => {
    await esops({cwd})

    await assertSnap({
      given: 'a minimal package with package.json',
      should: 'generate basic template in cwd',
      snap: getSortedFilePaths(cwd)
    })

    await assertSnap({
      given: 'a minimal package.json and included .gitignore',
      should: 'have updated .gitignore with generated paths',
      snap: getFileContents(path.join(cwd, '.gitignore'))
    })

    await assertSnap({
      given: 'a minimal package.json',
      should: 'have contents in package.json',
      snap: getJsonContents(path.join(cwd, 'package.json'))
    })
  })

  await withTempDir(__dirname, MOCK_STACKS['basic'], async cwd => {
    await esops({cwd})

    await assertSnap({
      given: 'a minimal package with no .gitignore',
      should: 'generate basic template in cwd',
      snap: getSortedFilePaths(cwd)
    })

    await assert({
      given: 'no .gitignore',
      should: 'not create a gitignore',
      expected: null,
      actual: getFileContents(path.join(cwd, '.gitignore'))
    })
  })

  await withTempDir(__dirname, MOCK_STACKS['basic-node-module'], async cwd => {
    await spawn(`npm`, ['install'], {cwd})
    await esops({cwd})
    await assertSnap({
      given: 'minimal stack from node module',
      should: 'generate basic template in cwd from node_module',
      snap: getSortedFilePaths(cwd)
    })
  })

  await withTempDir(__dirname, MOCK_STACKS['basic-bad-path'], async cwd => {
    const cleanError = cleanErrorString(cwd)
    const snap = cleanError(await Try(esops, {cwd}))
    await assertSnap({
      given: 'a bad path in the config',
      should: 'throw a friendly error',
      snap
    })
  })

  await withTempDir(__dirname, MOCK_STACKS['basic-bad-config'], async cwd => {
    const snap = cleanErrorString(cwd)(await Try(esops, {cwd}))
    await assertSnap({
      given: 'an invalid config object',
      should: 'throw a friendly error',
      snap
    })
  })

  await withTempDir(__dirname, MOCK_STACKS['basic-bad-config'], async cwd => {
    const snap = cleanErrorString(cwd)(await Try(esops, {cwd}))
    await assertSnap({
      given: 'non parseable',
      should: 'throw a friendly error',
      snap
    })
  })
})
