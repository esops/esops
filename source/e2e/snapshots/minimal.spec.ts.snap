module.exports[`Describe esops() minimal features | Given a minimal package with no extra files: should generate basic template in cwd`] = `

[
  ".vscode",
  ".vscode/settings.json",
  "esops.json",
  "infrastructure",
  "infrastructure/.vscode",
  "infrastructure/.vscode/settings.json",
  "infrastructure/src",
  "infrastructure/src/stores",
  "infrastructure/src/stores/stores-architecture.md",
  "infrastructure/tsconfig.json",
  "src",
  "src/stores",
  "src/stores/stores-architecture.md",
  "tsconfig.json"
]`

module.exports[`Describe esops() minimal features | Given an included .gitignore: should have updated .gitignore with generated file paths`] = `


### ESOPS AUTO GENERATED BEGIN ###
/.vscode/settings.json
/src/stores/stores-architecture.md
/tsconfig.json
### ESOPS AUTO GENERATED END ###
`

module.exports[`Describe esops() minimal features | Given an included .npmignore: should have updated .npmignore with generated file paths`] = `

/src/**
### ESOPS AUTO GENERATED BEGIN ###
/.vscode/settings.json
/src/stores/stores-architecture.md
/tsconfig.json
### ESOPS AUTO GENERATED END ###
`

module.exports[`Describe esops() minimal features | Given a minimal package with package.json: should generate basic template in cwd`] = `

[
  ".gitignore",
  ".vscode",
  ".vscode/settings.json",
  "infrastructure",
  "infrastructure/.vscode",
  "infrastructure/.vscode/settings.json",
  "infrastructure/src",
  "infrastructure/src/stores",
  "infrastructure/src/stores/stores-architecture.md",
  "infrastructure/tsconfig.json",
  "package.json",
  "src",
  "src/stores",
  "src/stores/stores-architecture.md",
  "tsconfig.json"
]`

module.exports[`Describe esops() minimal features | Given an included .gitignore: should have updated .gitignore with generated paths`] = `


### ESOPS AUTO GENERATED BEGIN ###
/.vscode/settings.json
/src/stores/stores-architecture.md
/tsconfig.json
### ESOPS AUTO GENERATED END ###
`

module.exports[`Describe esops() minimal features | Given a minimal package.json: should have contents in package.json`] = `

{
  "name": "esops-test",
  "esops": [
    "./infrastructure"
  ]
}`

module.exports[`Describe esops() minimal features | Given minimal stack from node module: should generate basic template in cwd from node_module`] = `

[
  ".gitignore",
  ".vscode",
  ".vscode/settings.json",
  "infrastructure",
  "infrastructure/basic",
  "infrastructure/basic/.vscode",
  "infrastructure/basic/.vscode/settings.json",
  "infrastructure/basic/src",
  "infrastructure/basic/src/stores",
  "infrastructure/basic/src/stores/stores-architecture.md",
  "infrastructure/basic/tsconfig.json",
  "infrastructure/package.json",
  "node_modules",
  "node_modules/@test",
  "node_modules/@test/infrastructure",
  "node_modules/@test/infrastructure/basic",
  "node_modules/@test/infrastructure/basic/.vscode",
  "node_modules/@test/infrastructure/basic/.vscode/settings.json",
  "node_modules/@test/infrastructure/basic/src",
  "node_modules/@test/infrastructure/basic/src/stores",
  "node_modules/@test/infrastructure/basic/src/stores/stores-architecture.md",
  "node_modules/@test/infrastructure/basic/tsconfig.json",
  "node_modules/@test/infrastructure/package.json",
  "package-lock.json",
  "package.json",
  "src",
  "src/stores",
  "src/stores/stores-architecture.md",
  "tsconfig.json"
]`

module.exports[`Describe esops() minimal features | Given no config found: should provide a friendly message on how to use esops`] = `

Error: No stack config found.

## Current Working Directory

/current/working/directory/

 ••• 
## How To
Esops works by adding a valid stack config to \`esops.json\` or \`package.json\`.

A valid stack config will be a path to a directory you want to copy from.

### Configure esops in \`package.json\` or \`esops.json\`

#### in \`esops.json\`

\`\`\`json
"node:@myorg/my-stack/stack"
\`\`\`

#### in \`package.json\`

\`{ "esops": "node:@myorg/my-stack/stack" }\`

### Resolution Types

- _Filesystem:_ \`'./infrastructure'\`
- _Node Module:_ \`'node:@myorg/my-stack/stack'\`

### Run \`esops\`

If global, just run \`esops\`. If local, use \`npx esops\`. Files will be generated, and if you have a \`.gitignore\`, it will be updated to include the generated files.

`

module.exports[`Describe esops() minimal features | Given a bad path in the config: should throw a friendly error`] = `

TypeError: Path \`./infrastructure\` not found.

## Current Working Directory

/current/working/directory/

 ••• 
## How To
Esops works by adding a valid stack config to \`esops.json\` or \`package.json\`.

A valid stack config will be a path to a directory you want to copy from.

### Configure esops in \`package.json\` or \`esops.json\`

#### in \`esops.json\`

\`\`\`json
"node:@myorg/my-stack/stack"
\`\`\`

#### in \`package.json\`

\`{ "esops": "node:@myorg/my-stack/stack" }\`

### Resolution Types

- _Filesystem:_ \`'./infrastructure'\`
- _Node Module:_ \`'node:@myorg/my-stack/stack'\`

### Run \`esops\`

If global, just run \`esops\`. If local, use \`npx esops\`. Files will be generated, and if you have a \`.gitignore\`, it will be updated to include the generated files.

`

module.exports[`Describe esops() minimal features | Given an invalid config object: should throw a friendly error`] = `

TypeError: Your stack definition is invalid.
Please add a a valid stack config to \`esops.json\` or \`package.json\`.

A valid stack config will be a path to a directory you want to copy from.

 ••• 
## How To
Esops works by adding a valid stack config to \`esops.json\` or \`package.json\`.

A valid stack config will be a path to a directory you want to copy from.

### Configure esops in \`package.json\` or \`esops.json\`

#### in \`esops.json\`

\`\`\`json
"node:@myorg/my-stack/stack"
\`\`\`

#### in \`package.json\`

\`{ "esops": "node:@myorg/my-stack/stack" }\`

### Resolution Types

- _Filesystem:_ \`'./infrastructure'\`
- _Node Module:_ \`'node:@myorg/my-stack/stack'\`

### Run \`esops\`

If global, just run \`esops\`. If local, use \`npx esops\`. Files will be generated, and if you have a \`.gitignore\`, it will be updated to include the generated files.

`

module.exports[`Describe esops() minimal features | Given non parseable: should throw a friendly error`] = `

TypeError: Your stack definition is invalid.
Please add a a valid stack config to \`esops.json\` or \`package.json\`.

A valid stack config will be a path to a directory you want to copy from.

 ••• 
## How To
Esops works by adding a valid stack config to \`esops.json\` or \`package.json\`.

A valid stack config will be a path to a directory you want to copy from.

### Configure esops in \`package.json\` or \`esops.json\`

#### in \`esops.json\`

\`\`\`json
"node:@myorg/my-stack/stack"
\`\`\`

#### in \`package.json\`

\`{ "esops": "node:@myorg/my-stack/stack" }\`

### Resolution Types

- _Filesystem:_ \`'./infrastructure'\`
- _Node Module:_ \`'node:@myorg/my-stack/stack'\`

### Run \`esops\`

If global, just run \`esops\`. If local, use \`npx esops\`. Files will be generated, and if you have a \`.gitignore\`, it will be updated to include the generated files.

`

module.exports[`Describe esops() minimal features | Given no to override: should not override files`] = `

{
  "has-been-overwritten": false
}
`

module.exports[`Describe esops() minimal features | Given yes to override: should override files`] = `

{
  "has-been-overwritten": true
}
`

module.exports[`Describe esops() minimal features | Given canceling prompt to override: should not override files`] = `

{
  "has-been-overwritten": false
}
`

